src/SBSpergel.cpp

```diff
@@ -0,0 +1,715 @@
+#ifdef DEBUGLOGGING
+            double x = x00;
+            double y = y00;
+            for (int j=0;j<intj0;++j) { x += dxy; y += dy; }
+            for (int i=0;i<inti0;++i) { x += dx; y += dyx; }
+            double r = sqrt(x*x+y*y);
+            dbg<<"Note: the original r value for this pixel had been "<<r<<std::endl;
+            dbg<<"xValue(r) = "<<_info->xValue(r)<<std::endl;
+            dbg<<"xValue(0) = "<<_info->xValue(0.)<<std::endl;
+#endif
+        }
+
+    }
+
+    void SBSpergel::SBSpergelImpl::fillKValue(tmv::MatrixView<std::complex<double> > val,
+                                            double x0, double dx, double dxy,
+                                            double y0, double dy, double dyx) const
+    {
+        dbg<<"SBSpergel fillKValue\n";
+        dbg<<"x = "<<x0<<" + ix * "<<dx<<" + iy * "<<dxy<<std::endl;
+        dbg<<"y = "<<y0<<" + ix * "<<dyx<<" + iy * "<<dy<<std::endl;
+        assert(val.stepi() == 1);
+        assert(val.canLinearize());
+        const int m = val.colsize();
+        const int n = val.rowsize();
+        typedef tmv::VIt<std::complex<double>,1,tmv::NonConj> It;
+
+        x0 *= _r0;
+        dx *= _r0;
+        dxy *= _r0;
+        y0 *= _r0;
+        dy *= _r0;
+        dyx *= _r0;
+
+        It valit = val.linearView().begin();
+        for (int j=0;j<n;++j,x0+=dxy,y0+=dy) {
+            double x = x0;
+            double y = y0;
+            for (int i=0;i<m;++i,x+=dx,y+=dyx) {
+                double ksq = x*x + y*y;
+                *valit++ = _flux * _info->kValue(ksq);
+            }
+        }
+    }
+
+    SpergelInfo::SpergelInfo(double nu, double trunc, const GSParamsPtr& gsparams) :
+        _nu(nu), _trunc(trunc), _gsparams(gsparams),
+        _gamma_nup1(boost::math::tgamma(_nu+1.0)),
+        _gamma_nup2(_gamma_nup1 * (_nu+1)),
+        _xnorm0((_nu > 0.) ? _gamma_nup1 / (2. * _nu) * std::pow(2., _nu) : INFINITY),
+        _truncated(_trunc > 0.),
+        _maxk(0.), _stepk(0.), _re(0.), _flux(0.),
+        _ft(Table<double,double>::spline)
+    {
+        dbg<<"Start SpergelInfo constructor for nu = "<<_nu<<std::endl;
+        dbg<<"trunc = "<<_trunc<<std::endl;
+
+        if (_nu < sbp::minimum_spergel_nu || _nu > sbp::maximum_spergel_nu)
+            throw SBError("Requested Spergel index out of range");
+    }
+
+    class SpergelIntegratedFlux
+    {
+    public:
+        SpergelIntegratedFlux(double nu, double gamma_nup2, double flux_frac=0.0)
+            : _nu(nu), _gamma_nup2(gamma_nup2),  _target(flux_frac) {}
+
+        double operator()(double u) const
+        // Return flux integrated up to radius `u` in units of r0, minus `flux_frac`
+        // (i.e., make a residual so this can be used to search for a target flux.
+        {
+            double fnup1 = std::pow(u / 2., _nu+1.)
+                * boost::math::cyl_bessel_k(_nu+1., u)
+                / _gamma_nup2;
+            double f = 1.0 - 2.0 * (1.+_nu)*fnup1;
+            return f - _target;
```

Comment:
I can't take credit for this result.  It's in Spergel (2010).  Here's the detailed comment I added though:

```
        // Return flux integrated up to radius `u` in units of r0, minus `flux_frac`
        // (i.e., make a residual so this can be used to search for a target flux.
        // This result is derived in Spergel (2010) eqn. 8 by going to Fourier space
        // and integrating by parts.
        // The key Bessel identities:
        // int(r J0(k r), r=0..R) = R J1(k R) / k
        // d[-J0(k R)]/dk = R J1(k R)
        // The definition of the radial surface brightness profile and Fourier transform:
        // Sigma_nu(r) = (r/2)^nu K_nu(r)/Gamma(nu+1)
        //             = int(k J0(k r) / (1+k^2)^(1+nu), k=0..inf)
        // and the main result:
        // F(R) = int(2 pi r Sigma(r), r=0..R)
        //      = int(r int(k J0(k r) / (1+k^2)^(1+nu), k=0..inf), r=0..R) // Do the r-integral
        //      = int(R J1(k R)/(1+k^2)^(1+nu), k=0..inf)
        // Now integrate by parts with
        //      u = 1/(1+k^2)^(1+nu)                 dv = R J1(k R) dk
        // =>  du = -2 k (1+nu)/(1+k^2)^(2+nu) dk     v = -J0(k R)
        // => F(R) = u v | k=0,inf - int(v du, k=0..inf)
        //         = (0 + 1) - 2 (1+nu) int(k J0(k R) / (1+k^2)^2+nu, k=0..inf)
        //         = 1 - 2 (1+nu) (R/2)^(nu+1) K_{nu+1}(R) / Gamma(nu+2)
```